src/vs/workbench/browser/parts/activitybar/activitybarPart.ts

```diff
@@ -477,8 +477,29 @@ export class ActivitybarPart extends Part implements IActivityBarService {
 		this.hideComposite(viewletId);
 	}
 
+	private updateActivity(viewlet: ViewletDescriptor, viewDescriptors: IViewDescriptorCollection): void {
+		const viewDescriptor = viewDescriptors.activeViewDescriptors[0];
+		const activity: IActivity = {
+			id: viewlet.id,
+			name: viewDescriptor.name,
+			cssClass: isString(viewDescriptor.containerIcon) ? viewDescriptor.containerIcon : undefined,
+			iconUrl: viewDescriptor.containerIcon instanceof URI ? viewDescriptor.containerIcon : undefined,
+			keybindingId: viewlet.keybindingId
+		};
+
+		const { activityAction, pinnedAction } = this.getCompositeActions(viewlet.id);
+		// if (activityAction instanceof PlaceHolderViewletActivityAction) {
+		activityAction.setActivity(activity);
+		// }
+
+		if (pinnedAction instanceof PlaceHolderToggleCompositePinnedAction) {
+			pinnedAction.setActivity(activity);
+		}
+	}
+
```

Comment:
Is not this change my viewlet icon every time I rearrange my views in the viewlet? I would recommend to have dynamic title and icons only for generated view containers and only gets updated when its associated view gets moved.

Reply:
I only included this code as a test as I thought you might have better expertise on the title/icon stuff. However, I still am not convinced we should treat built-in views differently that generated ones. Search without the search view is not Search to the user.

Reply:
I updated the heuristic to use the icon of the container if:
1) we have no icons from views (custom case for extension views)
2) we have a view with the same icon as the container (e.g. so long as search is in search, it will not change its icon)

This works in most cases but we still have an issue where the composite bar's representation (title in panel or icon in activity bar) is not accurate until the view is opened as it must use the static registration until the view is opened. This is similar to the issue we had with the debug console header in panel.

Reply:
Yeah. Will do some thinking and look into this.